16.0 create alt currency model

stephansainleger opened the pull request.

Commit:
[NEW] lcc_lokavaluto_app_connection: create res.alt.currency model

and associated views

Commit:
[IMP] lcc_lokavaluto_app_connection: link alt currency to wallet, credit and debit requests

Commit:
[IMP] lcc_lokavaluto_app_connection: transfer activate_automatic_topup

parameter in alt currency model

Commit:
[IMP] lcc_lokavaluto_app_connection: transfer commission_product_id

parameter in alt currency model

Commit:
[CLN] lcc_lokavaluto_app_connection: remove obsolete recipient filtering

parameter "allow_payments_only_to_companies" is not used, and should
be refactored in a more global recipient filtering rules.

Commit:
[IMP] lcc_lokavaluto_app_connection: add recipient search filtering on alt_currency

Commit:
[IMP] lcc_lokavaluto_app_connection: transfer _safe_wallet_partners()

from res.company to res.alt.currency models

Commit:
[IMP] lcc_lokavaluto_app_connection: transfer debit request cron

from res_company to res_alt_currency models

Commit:
[IMP] lcc_comchain_base: adapt add-on to res.alt.currency refactoring

[IMP] lcc_comchain_base: add unit tests on wallet model

[IMP] lcc_comchain_base: adapt wallet registration process

with alt currency name provided in the inputs parameters

Commit:
[IMP] lcc_cyclos_base: adapt add-on to res.alt.currency refactoring

[IMP] lcc_cyclos_base: add unit tests on wallet model

Commit:
[IMP] global: refactor wallet json data construction

Commit:
[FIX] lcc_lokavaluto_app_connection: correct partner_service tests

import PartnerChecktransaction class directly from code

Commit:
[FIX] lcc_credit_requests_from_contracts: typo

Commit:
[IMP] lcc_lokavaluto_app_connection: rename test file

Commit:
[IMP] lcc_lokavaluto_app_connection: res_partner.backend() now returns default foo currency wallets

for tests purposes

Commit:
[IMP] lcc_lokavaluto_app_connection: add partner services credit request tests

Commit:
[IMP] lcc_lokavaluto_app_connection: add partner service reconversions tests

Commit:
[WIP] lcc_lokavaluto_app_connection: add partner service search tests

Commit:
[IMP] global: add wallet ident and uri in wallet model

Commit:
[IMP] global: refactor _update_search_data()

factorize the function to remove specific code in comchain and cyclos addons

Commit:
[IMP] lcc_lokavaluto_app_connection: adapt credit requests services to new alt_currency model

Commit:
[IMP] lcc_lokavaluto_app_connection: adapt pending wallets service to new alt_currency model

Commit:
[IMP] lcc_lokavaluto_app_connection: adapt reconversion service to new alt_currency model

mondot commented on lcc_lokavaluto_app_connection/tests/test_partner_services_wallet_restriction_rules.py:
```diff
@@ -17,17 +26,18 @@ def setUp(self):
         )
 
         self.recipient_allowed = self.ResPartner.create({"name": "Allowed Recipient"})
         self.recipient_allowed_wallet = self.ResPartnerBackend.create(
```
Coucou, début de review : dans ces tests il faut ajouter "ident" pour tous les ResPartnerBackend.

mondot commented on lcc_lokavaluto_app_connection/tests/test_partner_services_wallet_restriction_rules.py:
```diff
@@ -139,7 +142,7 @@ def test_rule_allowing_all_senders(self):
             {
                 "name": "Test Rule",
                 "sender_wallet_domain": "[]",
```
Le test que j'avais écrit ne représente pas bien la réalité de ce qui se passe lorsqu'on omet de préciser un sender-wallet_domain depuis l'IHM. Il faut modifier comme cela

```suggestion
                "sender_wallet_domain": None,
```

Ce qui fait planter le test avec `ValueError: "'bool' object has no attribute 'strip'" while compiling`. Je propose une correction dans ta PR pour éviter les conflits.

mondot replied:
(vu avec Stephan) je propose une PR et il corrigera les conflits.

PR : https://github.com/Lokavaluto/lokavaluto-addons/pull/362/files

mondot commented on lcc_lokavaluto_app_connection/models/wallet.py:
```diff
@@ -83,6 +104,12 @@ def get_wallet_data(self):
+        By default, and if no wallet in self, only return alt_currency json data.
+        Need to be overrided by financial backend add-ons."""
+        return self.alt_currency_id.get_currency_json_data()
+
```
Je met mon commentaire ici mais ça concerne du code ligne 149, dans la fonction `get_first_matching_restriction_rule`

```python
        for rule in rules:
            if not rule.sender_wallet_domain:  # it means all senders are concerned
                return rule
            domain = safe_eval(rule.sender_wallet_domain) + [("id", "=", self.id)]
            if self.search(domain, limit=1):
                return rule
```

C'est pour faire passer le test `test_rule_allowing_all_senders`.

Commit:
[FIX] global: correct wallet ident management

Commit:
[IMP] lcc_lokavaluto_app_connection: concatenate same currency wallets in credential_backends service answer

Commit:
[IMP] global: add migration scripts

mondot commented:
Pair review avec Stephan : 

- [ ] Tester le fait que le code tourne si Monujo (ou n'importe qui d'autre) envoie des URIs.

Retours de Quentin (en solo) :

- [ ] Maj les fichiers `neutralize.sql` (actuellement ça maj les res_company) pour neutraliser les monnaies dans les DB neutralisées.
- [ ] Traductions
- [ ] Si possible, améliorer la couverture de tests de `lcc_lokavaluto_app_connection/models/res_alt_currency.py` (cf commentaire ci-dessous)
- [ ] Réparer import des data du manifest (remonter les fichiers security, cf commentaire)
- [ ] 1 petit test à corriger, cf commentaire

Jolie PR, elle fait plaisir :heart_eyes: good job

mondot commented on lcc_lokavaluto_app_connection/__manifest__.py:
```diff
@@ -36,11 +34,14 @@
         "data/numeric_data.xml",
-        "data/res_company_data.xml",
+        "data/res_alt_currency_data.xml",
+        "security/wallets_security.xml",
```
J'ai l'erreur suivante lors de l'initialisation de la DB en installant seulement lcc_lokavaluto_app_connection :

```
2025-10-29 14:08:17,372 939 ERROR odoo_test_res_alt_currency odoo.modules.registry: Failed to load registry
ValueError: External ID not found in the system: lcc_lokavaluto_app_connection.group_wallet_accounts_manager
odoo.tools.convert.ParseError: while parsing /opt/odoo/auto/dev/lokavaluto-addons/lcc_lokavaluto_app_connection/views/menu.xml:122, somewhere inside
<menuitem id="wallets_menu" parent="lcc_lokavaluto_app_connection.menu_digital_currencies" sequence="10" groups="lcc_lokavaluto_app_connection.group_wallet_accounts_manager" action="action_wallet_view"/>
```

J'ai dû remonter les fichiers security pour ne plus avoir l'erreur (l'ordre d'import est important).

mondot commented on lcc_lokavaluto_app_connection/tests/test_partner_services_reconversions.py:
```diff
@@ -0,0 +1,151 @@
+            result = service.reconversions(
+                params=reconversions_get_params
+            )
+            self.assertEqual(result[self.ident_debit_request_A], "received")
```
ce test fail pour moi

j'ai "draft" pour `result[self.ident_debit_request_A]` et `result[self.ident_debit_request_F]` (ce qui parait cohérent avec le fait que tu ne viens pas modifier le state au dessus).

mondot replied:
C'est la ligne
`invoice = self.env["account.move"].create(debit_invoice_data)` (dans la fonction create_debit_invoices de `lcc_lokavaluto_app_connection/models/debit_request.py`) qui lève cette exception :

UserError('No journal could be found in company YourCompany for any of those types: purchase')

mondot commented:
Le test coverage de `lcc_lokavaluto_app_connection` est pas mal :)

<img width="2560" height="1366" alt="image" src="https://github.com/user-attachments/assets/e1ef97a7-660e-433a-87d7-32322eaa0d01" />

Par contre je suis surpris de voir seulement 43% pour res.alt.currency.  
Comme il ne s'agit que de code nouveau dans cette PR, ça serait cool d'avoir une meilleure couverture de tests sur ce fichier.

À titre d'info je te mets les couvertures de tests de lcc_comchain_base et lcc_cyclos_base

<img width="1881" height="1057" alt="image" src="https://github.com/user-attachments/assets/d207f583-848b-4ccb-b51a-d1648786b080" />

<img width="1916" height="1255" alt="image" src="https://github.com/user-attachments/assets/aee57066-42a8-458f-8eec-3c95893364af" />

stephansainleger commented:
https://github.com/Lokavaluto/lokavaluto-addons/pull/353#issuecomment-3462018954
Merci pour le test coverage ! 
Effectivement on ne peut pas couvrir d'avantage res_alt_currency.py car l'une des fonctions principales dans ce fichier est le cron de récupération des reconversions : 
- ce cron nécessite que les modules pour Comchain ou pour Cyclos soient installés. Au pire on pourrait tester ces lignes via les autres modules :thinking: 
- OU alors il faut pouvoir simuler une réponse à la fonction `_retrieve_last_debit_transactions`, et je ne sais pas faire pour l'instant.
- mais quoiqu'il en soit, le test de ce cron ne peut pas se faire en l'état actuel du code, car il pourrait mettre des heures à tourner dans l'état actuel du code (je ne rentre pas dans les détails là). Un test limité pourrait être possible une fois que la PR https://github.com/Lokavaluto/lokavaluto-addons/pull/355 sera pushée aussi. Donc sujet à garder pour plus tard.

vaab commented on lcc_lokavaluto_app_connection/models/res_alt_currency.py:
```diff
@@ -0,0 +1,40 @@
+
+    ident = fields.Char("Currency Ident", tracking=True)
+    engine = fields.Selection(
+        [("foo", "Value defined only for testing purposes")],
```
hmm, not sure this is a good idea and why this is necessary.

vaab commented on lcc_lokavaluto_app_connection/tests/test_res_alt_currency.py:
```diff
@@ -0,0 +1,23 @@
+        self.currency_A_unit_product = self.env.ref(
+            "lcc_lokavaluto_app_connection.product_product_numeric_lcc"
+        ).sudo()
+        self.currency_A = self.ResAltCurrency.create(
```
This should be a test not the setUp. Using `.create()` on the model you create is a test. (and otherwise, you effectively shipped your mode without tests...).

vaab commented on lcc_lokavaluto_app_connection/models/res_alt_currency.py:
```diff
@@ -0,0 +1,40 @@
+    name = fields.Char("Name", tracking=True)
+    active = fields.Boolean(default=True, tracking=True)
+
+    ident = fields.Char("Currency Ident", tracking=True)
```
Not sure why this is not required from the start.

vaab commented on lcc_lokavaluto_app_connection/views/res_alt_currency.xml:
```diff
@@ -0,0 +1,60 @@
+        <field name="arch" type="xml">
+            <search string="Search Alternative Currencies">
+                <field name="name" />
+                <field name="technology" />
```
not (yet?) defined... or you meant `engine`

vaab reviewed (commented):
We'll finish in live.

Commit:
[IMP] global: factorize _get_backend_credentials() and _uptade_auth_data() code

move the code in a unique function get_partner_wallets_credentials()
in res.partner model

Commit:
[IMP] lcc_comchain_base: remove obsolete comchain_backend_id property

Commit:
[IMP] global: refactor backends() function

vaab committed:
[FIX] lcc_lokavaluto_app_connection: ruff pass
